### PR TITLE
[gh-1888] add account export module.

### LINKED ISSUE #1888: [CLI] Account command support export private key and mnemonic
Add the `rooch account export` command to export the account private key for one address or mnemonic for all addresses.

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

Add `account export` module and resolve #1888.

Testing various scenarios:

```zsh
> rooch account export       
Export succeeded with the mnemonic phrase [spot demise medal priority old rack lobster cover gasp just warfare essay]

> rooch account export -a rooch16469un6dceflr8026v0afv4t525d3y5s3u3g45r2kxxyj08twzdq4vtufw
Address to be exported: rooch16469un6dceflr8026v0afv4t525d3y5s3u3g45r2kxxyj08twzdq4vtufw(0xd5745e4f4dc653f19dead31fd4b2aba2a8d892908f228ad06ab18c493ceb709a)
Export succeeded with the encoded private key [roochsecretkey1qxmxtzkhftwa579390cj7v09e6xmsl6r4tgdy9q3rupwazaw065dj58kc75]

> rooch account export -a default
Address to be exported: rooch1xxzwqu47gsxf2s3max43t7tzneunljr7vpdy6d436tnw0e7zsp4q32jwdc(0x3184e072be440c95423be9ab15f9629e793fc87e605a4d36b1d2e6e7e7c2806a)
Export succeeded with the encoded private key [roochsecretkey1q8caw0k0m7uuydqectm9cums2ecz6fd4pccnut8vlkwwsyww6kxmxleuz8v]

> rooch account export -a ''     
Export succeeded with the mnemonic phrase [spot demise medal priority old rack lobster cover gasp just warfare essay]

> rooch account export -a ""
Export succeeded with the mnemonic phrase [spot demise medal priority old rack lobster cover gasp just warfare essay]

> rooch account export -a wwww
Invalid arguments: Invalid Rooch address String: Unbound named address: 'wwww'

> rooch account export -a rooch3xxzwqu47gsxf2s3max43t7tzneunljr7vpdy6d436tnw0e7zsp4q32jwdc
error: invalid value 'rooch3xxzwqu47gsxf2s3max43t7tzneunljr7vpdy6d436tnw0e7zsp4q32jwdc' for '--address <ADDRESS>': parsing failed

> rooch account export -a rooch1xwzwqu47gsxf2s3max43t7tzneunljr7vpdy6d436tnw0e7zsp4q32jwdc
error: invalid value 'rooch1xwzwqu47gsxf2s3max43t7tzneunljr7vpdy6d436tnw0e7zsp4q32jwdc' for '--address <ADDRESS>': no valid bech32 or bech32m checksum

> rooch account export -a rooch1zr6ut2z3kq7x6f587hezcx9kkdddk26aqy3cjag2lfvclx0ygnxs4ezhgq
Address to be exported: rooch1zr6ut2z3kq7x6f587hezcx9kkdddk26aqy3cjag2lfvclx0ygnxs4ezhgq(0x10f5c5a851b03c6d2687f5f22c18b6b35adb2b5d012389750afa598f99e444cd)
Error: Address rooch1zr6ut2z3kq7x6f587hezcx9kkdddk26aqy3cjag2lfvclx0ygnxs4ezhgq not found in the mnemonic data
```

- Closes #1888

### CODE REVIEW

File: crates/rooch-key/src/keystore/account_keystore.rs [hunk 1]
```diff
@@ -95,7 +95,7 @@ pub trait AccountKeystore {
             .addresses
             .iter()
             .position(|&target_address| target_address == address)
-            .unwrap() as u32;
+            .expect("Address not found") as u32;
```
Comment (jolestar) [inline crates/rooch-key/src/keystore/account_keystore.rs:98]: Raise an error?

File: crates/rooch-key/src/key_derive.rs [hunk 2]
```diff
@@ -21,6 +21,9 @@ pub const DERIVATION_PATH_PURPOSE_SCHNORR: u32 = 44;
 pub const DERIVATION_PATH_PURPOSE_ECDSA: u32 = 54;
 pub const DERIVATION_PATH_PURPOSE_SECP256R1: u32 = 74;
 pub const DERIVATION_PATH_PURPOSE_BIP86: u32 = 86;
+pub const SIGNATURE_SCHEME_FLAG_ED25519: u8 = 0x00;
+pub const SIGNATURE_SCHEME_FLAG_SECP256K1: u8 = 0x01;
```
Comment (jolestar) [inline crates/rooch-key/src/key_derive.rs:25]: Directly use the SignatureScheme in crypto.rs?

https://github.com/rooch-network/rooch/blob/477f8b7c4677d8c71ce2ed6f54a2eb341b488281/crates/rooch-types/src/crypto.rs#L53-L59

**Comment by jolestar** [comment]
@wow-sven Help review.

File: crates/rooch-key/src/keystore/account_keystore.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-key/src/keystore/account_keystore.rs]: Solved as above.

File: crates/rooch-key/src/key_derive.rs (hunk 2 above)
Comment (feliciss) [reply crates/rooch-key/src/key_derive.rs]: Resolved using SignatureScheme.

File: crates/rooch-key/src/keystore/account_keystore.rs [hunk 3]
```diff
@@ -69,6 +74,55 @@ pub trait AccountKeystore {
...
+                    address
+                ));
+            }
+        };
+        // get derivation path and seed
+        let derivation_path = generate_derivation_path(account_index)?;
+        let seed = derive_seed_from_mnemonic(Some(mnemonic_phrase), None)?; // assume word length is none
+                                                                            // derive 32 length secp256k1 key pair
+        let sk = derive_bitcoin_private_key_from_path(seed.as_bytes(), derivation_path)?;
+        let sk_bytes = sk.secret.as_ref();
```
Comment (jolestar) [inline crates/rooch-key/src/keystore/account_keystore.rs:113]: We should unlock the private key via password and then export it. So, `export_private_key` does not need the mnemonic argument.

Comment (feliciss) [reply crates/rooch-key/src/keystore/account_keystore.rs:113]: > We should unlock the private key via password and then export it.

I think every operations now are unlocked by password before exporting it:

```zsh
rooch account export -a rooch1ptyrgaz5gp5mjnunydg8lnalg7x5zqrhshr4c5g2qdy2u3wf5glqcm57m2

Enter the keystore password:
```

> So, `export_private_key` does not need the mnemonic argument.

The password is used for encryption not for decryption. In order to derive a key, a password is not needed, so there's no password for that.

The mnemonic argument is used for generating the seed and derive the secret key for function `derive_bitcoin_private_key_from_path`. Is there any other cases should I use to remove the mnemonic argument to derive the key?

Comment (feliciss) [reply crates/rooch-key/src/keystore/account_keystore.rs:113]: Okay. I now see a function `decrypt_with_type` was added.

**Review by wow-sven — APPROVED** [review]

File: crates/rooch-key/src/keystore/account_keystore.rs [hunk 4]
```diff
@@ -69,6 +71,29 @@ pub trait AccountKeystore {
...
+        // get 33 bytes flag and secret key
+        let mut priv_key_bytes = Vec::with_capacity(sk_bytes.len() + 1);
+        // supports secp256k1 signature scheme
+        priv_key_bytes.push(SignatureScheme::Secp256k1.flag());
+        priv_key_bytes.extend_from_slice(sk_bytes);
+        // init `roochsecretkey` as HRP
+        let hrp = Hrp::parse(ROOCH_SECRET_KEY_PREFIX)?;
+        // encode hrp and 33 bytes private key using bech32 method
+        let bech32_encoded = encode::<Bech32>(hrp, &priv_key_bytes)?;
+        Ok(bech32_encoded)
```
Comment (jolestar) [inline crates/rooch-key/src/keystore/account_keystore.rs:94]: Putting this function to RoochKeyPair is more suitable.

**Review by jolestar — APPROVED** [review]

**Comment by jolestar** [comment]
I merge this PR first.

**Pull request merged**